pkg/config/config.go

```diff
@@ -1,27 +1,150 @@
 package config
 
-type Group struct {
-	GrafanaAdmin bool  `json:"grafanaAdmin,omitempty"`
-	Orgs         []Org `json:"orgs"`
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+)
+
+type (
+	Config struct {
+		mu     sync.RWMutex
+		groups GroupsMap
+	}
+
+	// for marshalling purposes
+	config struct {
+		Groups GroupsMap `json:"groups"`
+	}
+
+	GroupsMap map[string]Group
+
+	Group struct {
+		GrafanaAdmin bool  `json:"grafanaAdmin,omitempty"`
+		Orgs         []Org `json:"orgs"`
+	}
+
+	Org struct {
+		ID   int64  `json:"id"`
+		Role string `json:"role"`
+	}
+)
+
+func New() *Config {
+	groupsMap := make(GroupsMap)
+
+	return &Config{
+		groups: groupsMap,
+	}
+}
+
+func (c *Config) SetGroup(name string, group Group) {
+	c.mu.Lock()
+	c.groups[name] = group
+	c.mu.Unlock()
 }
 
-type Groups map[string]Group
+func (c *Config) LoadGroup(name string) (Group, bool) {
+	c.mu.RLock()
+	result, ok := c.groups[name]
+	c.mu.RUnlock()
 
-type Org struct {
-	ID   int64  `json:"id"`
-	Role string `json:"role"`
+	return result, ok
 }
 
-// UserGroupsInConfig matches the user groups (from claims) that are
+func (c *Config) GroupsMap() GroupsMap {
+	return c.groups
+}
+
+func (c *Config) Load(file string) error {
+	data, err := os.ReadFile(file)
+	if err != nil {
+		return err
+	}
+
+	config := &config{}
+	if err = yaml.Unmarshal(data, config); err != nil {
+		return err
+	}
+
+	// Always update internal groups when loading from file
+	for gname, gconfig := range config.Groups {
+		c.SetGroup(gname, gconfig)
+	}
+
+	return nil
+}
+
+// ValidUserGroups matches the user groups (from claims) that are
 // present in config and returns a filtered set of Groups
-func ValidUserGroups(userGroups []string, groups Groups) Groups {
-	finalGroups := make(map[string]Group)
+func ValidUserGroups(userGroups []string, groupsMap GroupsMap) GroupsMap {
+	finalGroups := make(GroupsMap)
 
 	for _, userGroup := range userGroups {
-		if _, ok := groups[userGroup]; ok {
-			finalGroups[userGroup] = groups[userGroup]
+		if v, ok := groupsMap[userGroup]; ok {
```

Comment:
This should take []string, *config.Config and perform the look ups via Load(). Currently all lookups ignore the mutex during a reload.

Reply:
I changed the logic to read the groups from the Config object directly using LoadGroup